### PR TITLE
Pull request for branch "issue76" (#76)

### PR DESCRIPTION
**Pull request by mcjazzyfunky**
@thysultan 
As discussed: Support for "contextTypes" has not been implemented - you've said you take care of that on your own.
Could you please review my changes?
Please let me know in case I should change/add/fix something.
Thanks.

### CODE REVIEW

File: src/Core/Component.js [hunk 1]
```diff
@@ -465,5 +470,44 @@ function enqueueStateCallback (element, owner, callback) {
...
+function checkPropTypes(type, props) {
+	var propTypes = getPropTypes(type, props)
+
+	if (propTypes) {
+		var componentName = getDisplayName(type)
+
+		for (var propName in propTypes) {
+			var error = propTypes[propName](props, propName, componentName, 'prop', null)
+
+			if (error)
```
Comment (Zolmeister) [inline src/Core/Component.js]: could add a check for `Error` type for enforcing correctness

File: test/Context.spec.js [hunk 2]
```diff
@@ -819,5 +819,19 @@ describe('Context', () => {
...
+	}),
+
+	it('should write and read propTypes of context providers properly', () => {
+		const 
+			TestCtx1 = createContext(),
+			TestCtx2 = createContext(),
+			propTypes1 = {},
+			propTypes2 = {}
+
+		TestCtx1.Provider.propTypes = propTypes1
```
Comment (Zolmeister) [inline test/Context.spec.js]: consider using [Object.getOwnPropertyDescriptor()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyDescriptor) instead

File: test/Element.spec.js [hunk 3]
```diff
@@ -106,5 +106,84 @@ describe('Element', () => {
 
 		assert.deepEqual(element.props, {})
 		assert.lengthOf(element.children, 3)
+	}),
+
+	it('should validate propTypes in development envoironment and ignore in production environment', () => {
+		const
+			FunctionalComponent1 = ({ name }) => name,
```
Comment (Zolmeister) [inline test/Element.spec.js]: do we not lint bracket spacing? (`{name}` vs `{ name }`)

File: test/Element.spec.js [hunk 4]
```diff
@@ -106,5 +106,84 @@ describe('Element', () => {
...
+		FunctionalComponent2.displayName = 'SFComponent2'
+
+		ClassBasedComponent.propTypes = () => ({
+			name: () => null
+		})
+
+		if (process.env.NODE_ENV === 'development') {
+			assert.throws(task1, 'Validation failed for prop "name" of SFComponent1 (location: prop, fullPropName: null, props: {"name":"Jane Doe"}')
+			assert.doesNotThrow(task2)
+			assert.doesNotThrow(task3)
```
Comment (Zolmeister) [inline test/Element.spec.js]: we should also test that `ClassBasedComponent` throws

File: src/Core/Component.js (hunk 1 above)
Comment (mcjazzyfunky) [reply src/Core/Component.js]: @thysultan @Zolmeister 
Hey guys, what exactly do you want now? .. as you gave me different task descriptions regarding the error check ;-)

Option 1: if (error instanceof Error) // => as given in the first example snippet by @thysultan 
Option 2: if (error && typeof error.message === 'string') // => Quote: "Error-like to support cross realm instances"
Option 3: if (error) // => Quote:  "maybe this is enough(consider any truthy value as cause for exception)"

I personally prefer Option 1 as I cannot think of any use case where the "cross realm capability" regarding that error object is needed ... but maybe that's just due to lack of imagination ;-)

Comment (Zolmeister) [reply src/Core/Component.js]: This is what I meant:

```js
if (error) {
    if (error instanceof Error) {
        throw error
    else {
        throw new Error 'Return value must be of type `Error`'
    }
}
```

File: test/Context.spec.js (hunk 2 above)
Comment (mcjazzyfunky) [reply test/Context.spec.js]: @Zolmeister Mmmh, not really sure whether we really should use ```getPropertyDescriptor``` in the unit tests for this: The fact that the "propTypes" attribute is handled by getter and setter functions internally is just an implementation detail and therefore the aspect how the "propTypes" property is actually handled internally is not part of the public API (may change in future). As the unit tests do only test for the public API we should not use ```getOwnPropertyDescriptor``` IMHO....

Comment (Zolmeister) [reply test/Context.spec.js]: I see, this is testing `set` and `get` methods, not attribute writeability. disregard

**Comment by mcjazzyfunky** [comment]
@thysultan @Zolmeister Okay, I've fixed all issues discovered at @Zolmeister's review. The whole task is about making development with DIO more comfortable, so I pimped the error messages a bit to be also a bit "more comfortable".

**Comment by thysultan** [comment]
@mcjazzyfunky Lets move checkPropTypes out of Component into it's own file. i.e Types/Check?

File: src/Core/Context.js [hunk 5]
```diff
@@ -154,20 +154,53 @@ ContextConsumer[SharedSitePrototype] = ObjectCreate(Component[SharedSitePrototyp
...
+ContextProvider[SharedSiteDisplayName] = 'ContextProvider'
+ContextConsumer[SharedSiteDisplayName] = 'ContextConsumer'
+
 /**
  * @param {any} value
  * @return {{Provider, Consumer}}
  * @public
  */
 function createContext (value) {
+	var CustomContextProvider, Provider, Consumer
```
Comment (thysultan) [inline src/Core/Context.js]: Lets leave theses updates to `createContext` for another PR that also adds `contextTypes`.

File: src/Core/PropTypes.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+            + ' must be a function')
+      }
+
+      var error = validator(props, propName, componentName, 'prop', null)
+
+      if (error !== null) {
+        if (error instanceof Error) {
+          throw error
+        } else {
+          throw new Error(
```
Comment (thysultan) [inline src/Core/PropTypes.js]: We can leave this out to allow `return check[...] && Error(...)` pattern, that might return false.

File: src/Core/PropTypes.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  if (propTypes) {
+    var componentName = getDisplayName(type)
+
+    for (var propName in propTypes) {
+      var validator = propTypes[propName]
+
+      if (typeof validator !== 'function') {
+        throw new Error(
+          'Validator for propTypes.'
+            + propName
```
Comment (thysultan) [inline src/Core/PropTypes.js]: Just `Invalid propType validator` would suffice as error boundaries will generate a complete stack trace when this error propagates.

**Comment by mcjazzyfunky** [comment]
@thysultan 
I've commited all requested changes (and some more) to branch "issue76":

- Removed propTypes logic for ContextProvider (file "Context.js" is equal to "master" now)
- If a propType validator is not a function there will not be any special logic any longer => instead, a "normal" "is not a function" exception will be thrown
- Modified propTypes validator handling: Any result value is allowed now => true is okay, false will throw an error and everything else will also throw an error object if Boolean(result) === true
- Moved propTypes related unit tests to new file "PropTypes.spec.js"
- Fixed #81 and added two corresponding unit tests

File: test/Context.spec.js [hunk 8]
```diff
@@ -819,5 +819,19 @@ describe('Context', () => {
 
 		refs2.dispatchEvent(new Event('click'))
 		assert.html(container, `<button style="color: orange;">2</button><button>Colors!</button>`)
+	}),
+
+	it('should write and read propTypes of context providers properly', () => {
```
Comment (thysultan) [inline test/Context.spec.js]: We can add this to `PropTypes.spec.js` when we update context to support `propTypes`.

Comment (mcjazzyfunky) [reply test/Context.spec.js]: Uuups, thanks for the info ... I've completely forgotten that unit test :-( ... is removed

**Comment by thysultan** [comment]
LGTM, feel free to merge this. Thanks.

**Pull request merged**